File: crates/continuo-conductor/src/record.rs
//! The determinism harness's event log (milestone 2): tick fingerprints and
//! published messages, recorded as human-readable JSON lines.
//!
//! Recording taps the conductor's observation points — a `MonitorTransport`
//! callback for messages, and conductor callbacks for tick fingerprints,
//! membership changes, and over-budget steps — so the sim itself is
//! untouched by being recorded.
//!
//! Lines come in two categories: **expectations** are what the run did, and
//! a faithful re-run must produce them again; **observations** are what the
//! machine did, and a re-run is free to differ. Only the first kind is ever
//! compared. Every observation sits under [`LogEvent::Observed`], so which
//! category a line is in is read off its variant — see [`RecordedObservation`].
//!
//! A recorded log has two distinct consumers, near-opposite in how the
//! log's data flows relative to the simulation, one module each:
//!
//! - **Verification** ([`crate::Verifier`]): the log is an *expected-output
//!   ledger* and nothing flows into the sim. Every component re-runs live;
//!   each event is checked against the log as it happens, and the driving
//!   loop stops at the first divergence — which means "determinism is
//!   broken" (or the log was modified). Comparing two already-recorded
//!   logs lives there too.
//! - **Open-loop resimulation** ([`crate::PlaybackComponent`]): the log is an
//!   *input stimulus*. Selected recorded publishers are replaced by
//!   playback doubles that re-publish their recorded messages into the
//!   sim, while changed components run live against them. Nothing is
//!   compared; the new behavior diverging from the recording is the result
//!   being studied. (The played-back actors do not react to the live ones
//!   — that is what "open-loop" means.)

use std::path::Path;
use std::sync::{Arc, Mutex};

use continuo_core::{Message, SimTime, hash::hex_u64};
use serde::{Deserialize, Serialize};
use serde_json::value::RawValue;
use thiserror::Error;

use crate::config::ConductorConfig;
use crate::timing::OnTimeout;

/// Per-tick determinism fingerprint emitted by the conductor.
///
/// `tick_hash` covers everything the tick's stepped components did in
/// declaration order: paths, next-due times, published bytes, and (for
/// components implementing `state_bytes`) internal state. `world_hash`
/// chains tick hashes from the seeded initial value, so a single trailing
/// value fingerprints the entire run.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct TickFingerprint {
    pub tick: u64,
    pub sim_time: SimTime,
    #[serde(with = "hex_u64")]
    pub tick_hash: u64,
    #[serde(with = "hex_u64")]
    pub world_hash: u64,
}

/// A published message as recorded: paths flattened to strings, payload
/// embedded as raw JSON so the log stays readable.
#[derive(Debug, Serialize, Deserialize)]
pub struct RecordedMessage {
    pub time: SimTime,
    pub key: String,
    pub publisher: String,
    pub seq: u64,
    pub payload: Box<RawValue>,
}

/// A component admitted to the world, as recorded.
///
/// Deliberately *not* recorded: the sim time the join was applied at. It is
/// already implied by where this event sits between tick fingerprints, and
/// it is the part that may legitimately vary — once joins arrive over the
/// transport (milestone 7) the boundary that admits one depends on
/// delivery. What shapes the run is `first_due`, which the joiner declares,
/// so a run stays deterministically reproducible as long as that is
/// processed the same, whichever boundary the request landed on.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RecordedJoin {
    pub path: String,
    /// The instant the newcomer first steps.
    pub first_due: SimTime,
}

/// A component removed from the world, as recorded.
///
/// Carries the declared instant for the same reason a join carries
/// `first_due`: it is chosen by whoever asked, so it is stable however
/// early or late the request was made, and it is what decides where this
/// component's output stops. What is *not* recorded — here as on a join —
/// is the moment the request was applied, which says nothing extra and is
/// the part that varies with delivery.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RecordedLeave {
    pub path: String,
    /// The first instant the component did not step.
    pub leaves_at: SimTime,
}

/// A membership change, as handed to observers while it happens.
///
/// The log records *that* a component joined or left, not how to rebuild it
/// — a component cannot be reconstructed from bytes. Replaying a dynamic
/// run means re-running the scenario that issues the same joins and leaves,
/// and checking the resulting stream against the log.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum MembershipChange {
    Joined(RecordedJoin),
    Left(RecordedLeave),
}

/// A step that ran over the wall-clock budget its component declared.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct RecordedBudgetMiss {
    pub path: String,
    /// The instant the over-budget step ran at.
    pub sim_time: SimTime,
    /// What the step took, and what it was allowed, in wall-clock
    /// milliseconds.
    pub step_ms: f64,
    pub budget_ms: f64,
}

/// A component the conductor stopped waiting for, and what it did about it.
///
/// This is what says *why* a component left. The [`RecordedLeave`] that
/// follows a [`OnTimeout::Remove`] is deliberately identical to a scripted
/// one — the run behaved the same way, so replaying it by asking for that
/// leave at that instant must produce a matching log — and the reason lives
/// here, outside what verification compares. It is also the only record of a
/// [`OnTimeout::Halt`], which produces no leave at all: without this line a
/// halted run's log simply stops, saying nothing about why.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct RecordedTimeout {
    pub path: String,
    /// The instant whose step the conductor gave up waiting for.
    pub sim_time: SimTime,
    /// How long it waited, and how long it had declared it would, in
    /// wall-clock milliseconds.
    pub waited_ms: f64,
    pub timeout_ms: f64,
    /// What this did to the run — so a reader knows whether to expect a
    /// leave next or the log to end here.
    pub policy: OnTimeout,
}

/// Something the *machine* did rather than the run: one of the log's
/// **observations**.
///
/// The log holds two kinds of line. Every other kind is an **expectation**,
/// recording what the run did, and a faithful re-run must produce it again.
/// An observation is something noticed while the run happened that is worth
/// writing down, but that says nothing about whether the run was right.
///
/// They are noted centrally so they can all be read back from one file
/// afterwards, rather than from whichever process each step ran in.
/// Verification never compares them: another machine notices different
/// things, and treating that as a divergence would report two runs that
/// behaved identically as different.
///
/// Both members so far are wall-clock facts, because that is what milestones
/// 3 and 4 measure. Anything else of the same character belongs here rather
/// than as a new top-level event: the pacing overruns counted by
/// [`Conductor::overrun_reanchor_count`](crate::Conductor::overrun_reanchor_count)
/// are the obvious next one, being a run-level measurement that today exists
/// only as a counter that dies with the process.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum RecordedObservation {
    #[serde(rename = "budget")]
    BudgetMissed(RecordedBudgetMiss),
    #[serde(rename = "timeout")]
    TimedOut(RecordedTimeout),
}

/// One line of the log body, in emission order (messages of a tick precede
/// its fingerprint, since publishes happen during the steps; membership
/// changes sit between ticks, where they are applied).
///
/// Every observation nests under the one [`LogEvent::Observed`] variant
/// rather than getting a variant of its own, so that being an observation is
/// a structural fact instead of a list for each reader to keep in step with.
/// A new kind added to [`RecordedObservation`] is categorised correctly the
/// moment it exists; a new top-level variant would silently become an
/// expectation and start reporting false divergences.
#[derive(Debug, Serialize, Deserialize)]
pub enum LogEvent {
    #[serde(rename = "msg")]
    Msg(RecordedMessage),
    #[serde(rename = "tick")]
    Tick(TickFingerprint),
    #[serde(rename = "join")]
    Join(RecordedJoin),
    #[serde(rename = "leave")]
    Leave(RecordedLeave),
    /// Anything the machine did rather than the run — a log line reads
    /// `{"observed":{"budget":{...}}}`.
    #[serde(rename = "observed")]
    Observed(RecordedObservation),
}

impl LogEvent {
    /// The event's kind, for reporting a mismatch between two different
    /// kinds of event.
    pub(crate) fn kind(&self) -> &'static str {
        match self {
            LogEvent::Msg(_) => "msg",
            LogEvent::Tick(_) => "tick",
            LogEvent::Join(_) => "join",
            LogEvent::Leave(_) => "leave",
            LogEvent::Observed(RecordedObservation::BudgetMissed(_)) => "observed/budget",
            LogEvent::Observed(RecordedObservation::TimedOut(_)) => "observed/timeout",
        }
    }
}

impl From<RecordedObservation> for LogEvent {
    fn from(observation: RecordedObservation) -> Self {
        // Return the observation as the log line that records it.
        LogEvent::Observed(observation)
    }
}

impl From<MembershipChange> for LogEvent {
    fn from(change: MembershipChange) -> Self {
        // Return the change as the log line that records it.
        match change {
            MembershipChange::Joined(join) => LogEvent::Join(join),
            MembershipChange::Left(leave) => LogEvent::Leave(leave),
        }
    }
}

/// First line of the log: enough to re-create the run.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct LogHeader {
    pub version: u32,
    pub world_name: String,
    pub world_seed: u64,
}

#[derive(Debug, Error)]
pub enum RecordError {
    #[error("event log I/O failed: {0}")]
    Io(#[from] std::io::Error),
    #[error("event log line {line} is not valid: {source}")]
    Parse {
        line: usize,
        #[source]
        source: serde_json::Error,
    },
    #[error("event log is empty (missing header line)")]
    Empty,
}

/// A complete recorded run.
#[derive(Debug, Serialize, Deserialize)]
pub struct EventLog {
    pub header: LogHeader,
    pub events: Vec<LogEvent>,
}

impl EventLog {
    /// Serializes as JSON lines: header first, one event per line.
    pub fn to_jsonl(&self) -> String {
        let mut out = String::new();
        out.push_str(&serde_json::to_string(&self.header).expect("header serializes"));
        out.push('\n');
        for event in &self.events {
            out.push_str(&serde_json::to_string(event).expect("event serializes"));
            out.push('\n');
        }

        // Return the complete log as JSON-lines text.
        out
    }

    pub fn from_jsonl(text: &str) -> Result<Self, RecordError> {
        let mut lines = text
            .lines()
            .enumerate()
            .filter(|(_, l)| !l.trim().is_empty());
        let (_, header_line) = lines.next().ok_or(RecordError::Empty)?;
        let header: LogHeader = serde_json::from_str(header_line)
            .map_err(|source| RecordError::Parse { line: 1, source })?;
        let mut events = Vec::new();
        for (i, line) in lines {
            events.push(
                serde_json::from_str(line).map_err(|source| RecordError::Parse {
                    line: i + 1,
                    source,
                })?,
            );
        }

        // Return the parsed header and events.
        Ok(EventLog { header, events })
    }

    pub fn write_file(&self, path: impl AsRef<Path>) -> Result<(), RecordError> {
        Ok(std::fs::write(path, self.to_jsonl())?)
    }

    pub fn read_file(path: impl AsRef<Path>) -> Result<Self, RecordError> {
        Self::from_jsonl(&std::fs::read_to_string(path)?)
    }

    /// Latest world hash in the log, if any tick was recorded.
    pub fn final_world_hash(&self) -> Option<u64> {
        self.events.iter().rev().find_map(|e| match e {
            LogEvent::Tick(d) => Some(d.world_hash),
            _ => None,
        })
    }
}

/// Converts a live transport message into its recorded form.
pub(crate) fn recorded_message(m: &Message) -> RecordedMessage {
    let payload_text = std::str::from_utf8(&m.payload)
        .expect("payloads are serialized JSON, which is always valid UTF-8");
    let payload = RawValue::from_string(payload_text.to_string()).expect("payloads are valid JSON");

    // Return the message with paths flattened and payload embedded as raw
    // JSON.
    RecordedMessage {
        time: m.time,
        key: m.key.to_string(),
        publisher: m.publisher.to_string(),
        seq: m.seq,
        payload,
    }
}

/// Collects a run into an [`EventLog`] via two cloneable callbacks: attach
/// [`Recorder::message_callback`] to a `MonitorTransport` and
/// [`Recorder::tick_callback`] to `Conductor::set_tick_callback`, run, then call
/// [`Recorder::finish`].
#[derive(Clone)]
pub struct Recorder {
    inner: Arc<Mutex<EventLog>>,
}

impl Recorder {
    /// Starts a recording of the run `config` describes. The header comes
    /// from the same config the conductor runs with, so a log can never
    /// claim a world or seed the run did not use.
    pub fn new(config: &ConductorConfig) -> Self {
        // Return an empty log headed by this run's identity.
        Recorder {
            inner: Arc::new(Mutex::new(EventLog {
                header: LogHeader {
                    version: 1,
                    world_name: config.world_name.clone(),
                    world_seed: config.world_seed,
                },
                events: Vec::new(),
            })),
        }
    }

    pub fn message_callback(&self) -> impl FnMut(&Message) + Send + 'static {
        let inner = self.inner.clone();

        // Return the recording callback, holding its own handle to the
        // shared log.
        move |m: &Message| {
            inner
                .lock()
                .expect("recorder mutex is never poisoned")
                .events
                .push(LogEvent::Msg(recorded_message(m)));
        }
    }

    pub fn tick_callback(&self) -> impl FnMut(&TickFingerprint) + Send + 'static {
        let inner = self.inner.clone();

        // Return the recording callback, holding its own handle to the
        // shared log.
        move |d: &TickFingerprint| {
            inner
                .lock()
                .expect("recorder mutex is never poisoned")
                .events
                .push(LogEvent::Tick(*d));
        }
    }

    pub fn membership_callback(&self) -> impl FnMut(&MembershipChange) + Send + 'static {
        let inner = self.inner.clone();

        // Return the recording callback, holding its own handle to the
        // shared log.
        move |change: &MembershipChange| {
            inner
                .lock()
                .expect("recorder mutex is never poisoned")
                .events
                .push(change.clone().into());
        }
    }

    /// Records what the machine did — over-budget steps, and the timeouts
    /// that say why a component left or a run stopped — so a run's
    /// conditions end up in one file rather than in whichever process each
    /// step ran in. These lines are never compared; see [`RecordedObservation`].
    pub fn observation_callback(&self) -> impl FnMut(&RecordedObservation) + Send + 'static {
        let inner = self.inner.clone();

        // Return the recording callback, holding its own handle to the
        // shared log.
        move |observation: &RecordedObservation| {
            inner
                .lock()
                .expect("recorder mutex is never poisoned")
                .events
                .push(observation.clone().into());
        }
    }

    /// Takes the collected log. Other callback clones may still exist (inside a
    /// conductor); this snapshots the current contents.
    pub fn finish(&self) -> EventLog {
        let guard = self.inner.lock().expect("recorder mutex is never poisoned");

        // Return a snapshot of everything recorded so far.
        EventLog {
            header: guard.header.clone(),
            events: serde_json::from_str(
                &serde_json::to_string(&guard.events).expect("events serialize"),
            )
            .expect("events round-trip"),
        }
    }
}

File: python/continuo_viz/record.py
"""What a viewer reads, independent of where it came from.

A recorded log line and a live Zenoh sample carry the same information in two
arrangements, so both are parsed into the types here and nothing downstream
knows which was attached:

- a log line is ``{"msg": {time, key, publisher, seq, payload}}``, complete in
  itself because lines of every kind share one file
- a Zenoh sample is the payload bytes with ``{time, key, publisher, seq}``
  attached, because the payload already travelled as the payload and sending
  it twice would double every byte on the wire

The ``key`` in both cases is the key the component *published* on, not the
viewer side channel it was relayed onto, which is what lets one parser name
actors the same way for a live run and a replay.
"""

from __future__ import annotations

import json
import math
import re
from dataclasses import dataclass
from typing import Any

# The log header's `version` field. A log written by an older or newer
# continuo is not silently half-understood: parsing refuses it instead.
LOG_VERSION = 1


class UnsupportedLogVersion(Exception):
    """A log declares a format this viewer does not know how to read."""


@dataclass(frozen=True, slots=True)
class Message:
    """One published message, with the provenance a raw payload lacks."""

    time: float
    key: str
    publisher: str
    seq: int
    payload: dict[str, Any]


@dataclass(frozen=True, slots=True)
class Join:
    """A component admitted to the world."""

    path: str
    first_due: float


@dataclass(frozen=True, slots=True)
class Leave:
    """A component removed from the world.

    The exact event a viewer needs in order to stop drawing something. Without
    it the only options are a staleness timer, which cannot tell a departed
    actor from a stalled simulation, or drawing ghosts forever.
    """

    path: str
    leaves_at: float


Event = Message | Join | Leave

# `continuo/{world}/actor/{name}/{signal}`. Anchored at both ends so a longer
# key cannot match by accident.
_ACTOR_KEY = re.compile(r"^continuo/[^/]+/actor/(?P<actor>[^/]+)/(?P<signal>[^/]+)$")


def actor_signal(key: str) -> tuple[str, str] | None:
    """Splits an actor key into its actor name and signal.

    Returns ``None`` for anything else, which is how world-level traffic and
    conductor notifications are ignored without listing them.
    """
    match = _ACTOR_KEY.match(key)
    if match is None:
        return None
    return match["actor"], match["signal"]


@dataclass(frozen=True, slots=True)
class Pose:
    """A pose flattened to what a top-down view can draw.

    The simulation works in three dimensions and a full quaternion. A plan view
    needs two of the axes and one angle, so the conversion happens once here
    rather than in the renderer every frame.
    """

    x: float
    y: float
    z: float
    yaw: float
    """Heading in radians, counter-clockwise from the +x axis."""


def pose_from_payload(payload: dict[str, Any]) -> Pose | None:
    """Reads a pose payload, or ``None`` if it is not one.

    Returning ``None`` rather than raising keeps an unexpected payload on a
    pose key from ending a live session. A viewer that stops at the first
    surprise is worse than one that draws what it understood.
    """
    position = payload.get("position")
    orientation = payload.get("orientation")
    if not isinstance(position, dict) or not isinstance(orientation, dict):
        return None
    try:
        x = float(position["x"])
        y = float(position["y"])
        z = float(position["z"])
        w = float(orientation["w"])
        qx = float(orientation["x"])
        qy = float(orientation["y"])
        qz = float(orientation["z"])
    except (KeyError, TypeError, ValueError):
        return None

    # Yaw about +z, the only rotation a plan view can show. The standard
    # extraction, and stable for the near-level orientations a road produces.
    yaw = math.atan2(2.0 * (w * qz + qx * qy), 1.0 - 2.0 * (qy * qy + qz * qz))
    return Pose(x=x, y=y, z=z, yaw=yaw)


def check_log_version(header: dict[str, Any]) -> None:
    """Raises unless a log header declares a format this viewer reads."""
    version = header.get("version")
    if version != LOG_VERSION:
        raise UnsupportedLogVersion(
            f"log declares version {version!r}, this viewer reads {LOG_VERSION}"
        )


def event_from_log_line(line: str) -> Event | None:
    """Parses one line of a recorded log.

    Returns ``None`` for a line the viewer has no use for, which is most of
    them: tick fingerprints outnumber everything, and observations describe
    what the machine did rather than what the world did. Skipping by returning
    ``None`` rather than by filtering on a list of known kinds means a log kind
    added later is ignored instead of crashing a viewer that predates it.
    """
    stripped = line.strip()
    if not stripped:
        return None
    event = json.loads(stripped)
    if not isinstance(event, dict):
        return None

    if (msg := event.get("msg")) is not None:
        return Message(
            time=float(msg["time"]),
            key=str(msg["key"]),
            publisher=str(msg["publisher"]),
            seq=int(msg["seq"]),
            payload=msg["payload"],
        )
    if (join := event.get("join")) is not None:
        return Join(path=str(join["path"]), first_due=float(join["first_due"]))
    if (leave := event.get("leave")) is not None:
        return Leave(path=str(leave["path"]), leaves_at=float(leave["leaves_at"]))
    return None


def event_from_sample(payload: bytes, attachment: bytes | None) -> Event | None:
    """Parses one live sample into the same event a log line would give.

    The two halves are recombined here: a message arrives as payload bytes with
    provenance attached, so this is where they become the single record a log
    line already is. A sample with no attachment is a conductor notification,
    whose payload is a complete log line and needs nothing alongside it.
    """
    if attachment is None:
        return event_from_log_line(payload.decode("utf-8"))

    meta = json.loads(attachment.decode("utf-8"))
    return Message(
        time=float(meta["time"]),
        key=str(meta["key"]),
        publisher=str(meta["publisher"]),
        seq=int(meta["seq"]),
        payload=json.loads(payload.decode("utf-8")),
    )
